streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java

```diff
@@ -0,0 +1,295 @@
+            findClientsOnUsedClientTagDimensions(
+                activeTaskClient,
+                countOfUsedClients,
+                rackAwareAssignmentTags,
+                clientStates,
+                tagEntryToClients,
+                tagKeyToValues
+            )
+        );
+
+        while (numRemainingStandbys > 0) {
+            final UUID clientOnUnusedTagDimensions = standbyTaskClientsByTaskLoad.poll(
+                activeTaskId, uuid -> !clientsOnAlreadyUsedTagDimensions.contains(uuid)
+            );
+
+            if (clientOnUnusedTagDimensions == null) {
+                break;
+            }
+
+            clientStates.get(clientOnUnusedTagDimensions).assignStandby(activeTaskId);
+
+            countOfUsedClients++;
+
+            clientsOnAlreadyUsedTagDimensions.addAll(
+                findClientsOnUsedClientTagDimensions(
+                    clientOnUnusedTagDimensions,
+                    countOfUsedClients,
+                    rackAwareAssignmentTags,
+                    clientStates,
+                    tagEntryToClients,
+                    tagKeyToValues
+                )
+            );
+
+            numRemainingStandbys--;
+        }
+
+        return numRemainingStandbys;
+    }
+
+    private static Set<UUID> findClientsOnUsedClientTagDimensions(final UUID usedClient,
+                                                                  final int countOfUsedClients,
+                                                                  final Set<String> rackAwareAssignmentTags,
+                                                                  final Map<UUID, ClientState> clientStates,
+                                                                  final Map<TagEntry, Set<UUID>> tagEntryToClients,
+                                                                  final Map<String, Set<String>> tagKeyToValues) {
+        final Set<UUID> filteredClients = new HashSet<>();
+
+        final Map<String, String> usedClientTags = clientStates.get(usedClient).clientTags();
+
+        for (final Entry<String, String> usedClientTagEntry : usedClientTags.entrySet()) {
+            final String tagKey = usedClientTagEntry.getKey();
+
+            if (!rackAwareAssignmentTags.contains(tagKey)) {
+                log.warn("Client tag with key [{}] will be ignored when computing rack aware standby " +
+                         "task assignment because it is not part of the configured rack awareness [{}].",
+                         tagKey, rackAwareAssignmentTags);
+                continue;
+            }
+
+            final Set<String> allTagValues = tagKeyToValues.get(tagKey);
+            final String tagValue = usedClientTagEntry.getValue();
+
+            // If we have used more clients than all the tag's unique values,
+            // we can't filter out clients located on that tag.
+            if (allTagValues.size() <= countOfUsedClients) {
+                continue;
```

Comment:
Sorry, I didn't understand the reason why we can't filter out clients located on that tag when `allTagValues.size() <= countOfUsedClients`. Could you help explain to me? Thanks.

Reply:
Consider the following example

```
# Kafka Streams Client 1
client.tag.zone: eu-central-1a
client.tag.cluster: k8s-cluster1
rack.aware.assignment.tags: zone,cluster

# Kafka Streams Client 2
client.tag.zone: eu-central-1b
client.tag.cluster: k8s-cluster1
rack.aware.assignment.tags: zone,cluster
 
# Kafka Streams Client 3
client.tag.zone: eu-central-1c
client.tag.cluster: k8s-cluster1
rack.aware.assignment.tags: zone,cluster

# Kafka Streams Client 4
client.tag.zone: eu-central-1a
client.tag.cluster: k8s-cluster2
rack.aware.assignment.tags: zone,cluster
 
# Kafka Streams Client 5
client.tag.zone: eu-central-1b
client.tag.cluster: k8s-cluster2
rack.aware.assignment.tags: zone,cluster
 
# Kafka Streams Client 6
client.tag.zone: eu-central-1c
client.tag.cluster: k8s-cluster2
rack.aware.assignment.tags: zone,cluster
```

With the above we have following number of unique tag values:

```
zone: 3 (eu-central-1a, eu-central-1b, eu-central-1c)
cluster: 2 (k8s-cluster1, k8s-cluster2)
```

Now lets say we have standby replica count as `2` and we want to active task is located on `Client 1`

`usedClients=1` (because of the active task)

### 1s standby assignment

During 1st standby takes assignment, we will exclude clients on following dimensions:

```
zone: eu-central-1a
cluster: k8s-cluster1
```

Used clients will get incremented since we can allocate the client on different `zone` and `cluster`.
`usedClients=2`

### 2nd standby assignment

We will have to exclude `zone: eu-central-1a and (eu-central-1b || eu-central-1c)` tag values. We can do that, because
after we exclude clients on the new tag, we still have clients on the one free tag value we can assign the next standby
to.

We can't exclude `cluster` because we have already used two clients, and we just have two unique values for
the `cluster`
tag, so it's impossible to get the ideal distribution with this configuration and number of instances. So we can only
achieve partial distribution.

So idea of this check is to ignore tags where we have less unique values than the clients we have already used. If we
don't have this check, for the 2nd standby task assignment we would have excluded all the clients located
on `k8s-cluster1` and `k8s-cluster2`, and there wouldn't be any client left to assign the standby task to. We would fall
back to the least loaded client, BUT there will be no guarantee that least loaded client assignment would honor partial
rack awareness.

Hope this makes sense.

Reply:
I added `shouldDoThePartialRackAwareness` test to verify this behaviour.

Reply:
Added more explanation in `shouldDoThePartialRackAwareness` test.